xbrowse_server/phenotips/views.py

```diff
@@ -1,90 +1,187 @@
+  project_phenotips_uname,project_phenotips_pwd = get_uname_pwd_for_project(project_name)
+  try:
+    result = do_authenticated_call_to_phenotips(__aggregate_url_parameters(request),project_phenotips_uname,project_phenotips_pwd)
+    return __add_back_phenotips_headers_response(result)
+  except Exception as e:
+    print 'proxy get error:',e
+    logger.error('phenotips.views:'+str(e))
+    raise Http404
+
+
+#given a request object,and base URL aggregates and returns a reconstructed URL
+def __aggregate_url_parameters(request):
+  '''given a request object, aggregates and returns parameters'''
+  counter=0
+  url=settings.PHENOPTIPS_HOST_NAME+request.path + '?'
+  for param,val in request.GET.iteritems():
+    url += param + '=' + val
+    if counter < len(request.GET)-1:
+      url += '&'
+    counter+=1
+  return url
+
+#do a POST as a proxy for Phenotips
+#exempting csrf here since phenotips doesn't have this support
+@csrf_exempt
+def proxy_post(request):
+  '''to act as a proxy  '''
+  try:    
+    if len(request.POST) != 0 and request.POST.has_key('PhenoTips.PatientClass_0_external_id'):
+      project_name = request.session['current_project_name']
+      uname,pwd = get_uname_pwd_for_project(project_name)
+      __process_sync_request_helper(request.POST['PhenoTips.PatientClass_0_external_id'],
+                                    uname,
+                                    pwd,
+                                    request.user.username,
+                                    project_name)
+    #re-construct proxy-ed URL again
+    url=settings.PHENOPTIPS_HOST_NAME+request.path
+    project_name = request.session['current_project_name']
+    uname,pwd = get_uname_pwd_for_project(project_name)
+    resp = requests.post(url, data=request.POST, auth=(uname,pwd))
+    response = HttpResponse(resp.text)
+    for k,v in resp.headers.iteritems():
+      response[k]=v
+    return response
+  except Exception as e:
+    print 'proxy post error:',e
+    logger.error('phenotips.views:'+str(e))
+    raise Http404
+  
+  
 #process a synchronization between xbrowse and phenotips
-def __process_sync_request_helper(uname,pwd,eid):
+def __process_sync_request_helper(int_id,uname,pwd,xbrowse_username,project_name):
   '''sync data of this user between xbrowse and phenotips'''  
   try:
     #first get the newest data via API call
-    url= os.path.join(settings.PHENOPTIPS_HOST_NAME,'bin/get/PhenoTips/ExportPatient?eid='+eid)
-    password_mgr = urllib2.HTTPPasswordMgrWithDefaultRealm()
-    request = urllib2.Request(url)
-    base64string = base64.encodestring('%s:%s' % (uname, pwd)).replace('\n', '')
-    request.add_header("Authorization", "Basic %s" % base64string)   
-    result = urllib2.urlopen(request)   
-    print result.read()
+    url= os.path.join(settings.PHENOPTIPS_HOST_NAME,'bin/get/PhenoTips/ExportPatient?eid='+int_id)
+    result = do_authenticated_call_to_phenotips(url,uname,pwd)
+    updated_patient_record=eval(result.read())
```

Comment:
Yeah, I agree, I am not a fan of eval either. I always use JSON parsing, but unfortunately that did't work well here due JSON enoding on the phenotips side sending back backslashes (http://stackoverflow.com/questions/10314715/why-is-json-encode-adding-backslashes). I took out the JSON code and evaluated into a dict to make it work. Since this JSON is coming from the server side from Phenotips (and not from the external internet), I trusted it for during testing, but I agree this might be a hole where people can save something bad.

I put some time into this again and found a way around it via,
updated_patient_record=json.dumps(json.JSONDecoder().decode(result.read()))

This solved the issue. Took out eval, put above in, committed.